### PR TITLE
Report unprocessable entity

### LINKED ISSUE #318: Generic error message displayed when providing invalid chart name for app deployment
When trying to deploy an application and providing an invalid name - such as starting with upper case - kubeapps shows a very generic and not informative error:

image

It would be beneficial if there is validation for some of the fields and/or the output from deployment error is parsed and some errors are shown with their message.

### PR DESCRIPTION
**Pull request by andresmgot**
Fixes #318

 - In the Dashboard we now validate that the name given by the user passes the regex expected. If that is not the case, it shows a red border and hides the "Submit" button. TODO: Show a warning message explaining why the name is not valid:

screen shot 2018-08-10 at 11 54 58

In case the basic validation doesn't catch the error:

 - The Tiller proxy returns now a 422 error (Unprocessable entity) if it fails to deploy a release.
 - The dashboard shows the raw error that Tiller returns for the user to debug:

screen shot 2018-08-09 at 16 34 21

Apart from that the `AppRepositoryButton` now catches as well name conflicts and unprocessable entities as well.

### CODE REVIEW

**Comment by migmartri** [comment]
Thanks @andresmgot 

Left some comments, I'd also split this PR in two. One for the repository client side validation and another one for the unprocessable entity.

File: dashboard/src/components/DeploymentForm/DeploymentForm.tsx [hunk 1]
```diff
@@ -163,9 +167,13 @@ class DeploymentForm extends React.Component<IDeploymentFormProps, IDeploymentFo
                 />
               </div>
               <div>
-                <button className="button button-primary" type="submit">
-                  Submit
-                </button>
+                {this.state.releaseNameError ? (
```
[the remark this answers is not in the document]
Comment (andresmgot) [reply dashboard/src/components/DeploymentForm/DeploymentForm.tsx]: I didn't know that, thanks!

File: dashboard/src/components/DeploymentForm/DeploymentForm.tsx [hunk 2]
```diff
@@ -195,6 +203,9 @@ class DeploymentForm extends React.Component<IDeploymentFormProps, IDeploymentFo
   };
 
   public handleReleaseNameChange = (e: React.FormEvent<HTMLInputElement>) => {
+    this.state.releaseNameError = !e.currentTarget.value.match(
```
[the remark this answers is not in the document]
Comment (andresmgot) [reply dashboard/src/components/DeploymentForm/DeploymentForm.tsx]: Again I didn't know that, thanks :)

**Comment by andresmgot** [comment]
> One for the repository client side validation and another one for the unprocessable entity.

@migmartri you mean to do a standalone PR for the `form` validation, isn't it? That is now a two line change. Do you still want me to split this PR in two?

File: cmd/tiller-proxy/handler.go [hunk 3]
```diff
@@ -108,14 +109,25 @@ func isForbidden(err error) bool {
...
+	re := regexp.MustCompile(`release\s+[^\s]\s+failed`)
+	return re.MatchString(err.Error())
+}
+
 func errorCode(err error) int {
-	errCode := http.StatusInternalServerError
+	return errorCodeWithDefault(err, http.StatusInternalServerError)
+}
+
+func errorCodeWithDefault(err error, defaultCode int) int {
```
Comment (migmartri) [inline cmd/tiller-proxy/handler.go]: Good solution!

File: dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx [hunk 4]
```diff
@@ -163,6 +171,8 @@ export class AppRepoAddButton extends React.Component<
             action={`create AppRepository "${name}"`}
           />
         );
+      case UnprocessableEntity:
+        return <UnexpectedErrorAlert text={error && error.message} raw={true} />;
```
Comment (migmartri) [inline dashboard/src/components/Config/AppRepoList/AppRepoButton.tsx]: Great :)

File: dashboard/src/shared/types.ts [hunk 5]
```diff
@@ -36,6 +36,13 @@ export class AppConflict extends Error {
   }
 }
 
+export class UnprocessableEntity extends Error {
+  constructor(message?: string) {
+    super(message);
+    Object.setPrototypeOf(this, new.target.prototype);
```
Comment (migmartri) [inline dashboard/src/shared/types.ts]: I am not that proficient in JS. Why is ` Object.setPrototypeOf` this required, that's its purpose?

Comment (prydonius) [reply dashboard/src/shared/types.ts]: So I stole this from https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget, honestly I don't fully understand it myself, but we _should_ understand it. @Angelmmiguel could you maybe look into if/why we need this?

Comment (andresmgot) [reply dashboard/src/shared/types.ts]: I see this is going to be tackled in #489

**Comment by andresmgot** [comment]
@migmartri I agree with your comments. I tend to accumulate unrelated changes in PRs and I am aware of the downsides of that. I'll try to have that in mind in future PRs. 

Regarding the unit tests for the HTTP handlers of the `tiller-proxy` I am going to open a new PR for that so we avoid more noise in this accepted PR, I hope that's okay. Thanks for the reviews!

**Pull request merged**